### ISSUE TITLE
Stream image rendering

### ISSUE THREAD

**Issue by CosmicHorrorDev** [trimental/inlyne#78]
Opening this since @trimental is much more knowledgeable of the GPU related parts

After #74 the final piece  of the lower peak memory usage puzzle is preventing decoding the entire compressed image blob in memory to render. The reasonable solution here, to me at least, seems to be chunking the image horizontally to render _N_ rows instead of rendering the whole thing at once. This would lower the memory usage since we can control the number of rows used to use a smaller buffer. Does this seem workable @trimental?

**Comment by trimental** [comment]
Right now the most efficient way to upload our image data from normal cpu ram memory to gpu texture memory is to use [write_texture](https://docs.rs/wgpu/latest/wgpu/struct.Queue.html#method.write_texture) which is stores the data in staging memory and submits it to the gpu together with other queued gpu commands. 

Even if we were to use multiple calls to `write_texture` so they were row by row, they would still be stored in staging memory until the submit I believe so there would be not point. It might be possible to stream an upload to a gpu buffer and then copy to a gpu texture but this would have performance hits.

I think the only proper way to get even lower memory usage would be to used a compressed texture format and write that to the gpu.

**Comment by CosmicHorrorDev** [comment]
Using compressed textures sounds exactly like what I would want :eyes:

I'm not very familiar with stuff in the GPU space. Got a link or anything I can look at?

**Comment by trimental** [comment]
Yep sure [here](https://github.com/gfx-rs/wgpu/blob/628a95cd1c27abce4d4ff2ffe4558cc4ef1e7fd3/wgpu/examples/skybox/main.rs#L271). I'm not sure how much memory these compressed textures save though or how to rewrite images into them.
